lib/rust/data-structures/src/counter.rs

```diff
@@ -0,0 +1,134 @@
+//! # Counter
+//!
+//! Counter type. Uses an internal global value to ensure every instance created has a different
+//! value.
+
+use crate::prelude::*;
+
+
+
+// ===============
+// === Counter ===
+// ===============
+
+/// Implements a globally-unique counter.
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq, Ord, PartialOrd)]
+pub struct Counter {
+    value: std::num::NonZeroU64,
```

Comment:
out of curiosity, why don't we allow zero-ids? The explanation for this design should be provided in the docs.

Reply:
The optimizer uses non-zero-values to avoid the space overhead of an outer `Option`. Will add to docs.